File: DESC_MATLAB/src/specprocess_sourcepara.m
function  para_EGF = specprocess_sourcepara(rdir,freqfit,evspec,...
    imethod,iwave, EGF,min_nspec,isavefig)
% this is only for constant EGF!
% individual pair will  use a different program to use individual stations

if (~exist(rdir,'dir'))
    mkdir(rdir);
end
if (rdir(end:end)~='/') rdir(end+1) = '/'; end

freq = evspec(1).freq;
frqlim=[0.5 50];
flow = [0.5 5];
fhigh= [20 30];
nlow = find(freq>=flow(1) & freq<=flow(2));
nhigh= find(freq>=fhigh(1) & freq<=fhigh(2));
nfit = find(freq>=freqfit(1) & freq<=freqfit(2));

fid4 = fopen(strcat(rdir,'result-constant-egf.txt'),'w');
format4 = [' %5d','%8.2f','%8.2f','%15d',...
    '%15.2f','%15.2f','%15.2e','%10.3e','%10.3e',...
    '%11.3e','%10.3e','%10.3e','%10.3e','%10.3e','%10.3e %10.3e' ...
    ' %15.2f %15.2f \n'];

header4 = ['    i1   magest magcat              id1  ',...
    '     o0             fc1          r    delsig    asig    ',...
    '     G          Es-all   Es-model   ',...
    'Es-obs   Es-1       Es-2   error fcbound1 fcbound2 \n'];
fprintf(fid4,header4);

im = imethod;
%if (imethod == 2) im = 1; end
%if (imethod == 3) im = 5; end
nq = length(evspec);

nqsolve = 0;
for iq = 1:nq
    if (evspec(iq).nspec >= min_nspec) % then will solve using constant-EGF
        dspecc = evspec(iq).spec - EGF;

        % compute a few quality parameters
        if ((max(dspecc)-min(dspecc))>1e-2) % exist
            iexist = 1;
            nqsolve = nqsolve+1;
            [fc,fc2,o0,Es,delsig,asig,r,G,err,dspecfit,iflag,ftest,etest,fcb1,fcb2] = ...
                sourcepara (dspecc,freq,nfit,nlow,im,evspec(iq).qmomest,iwave);
            lhratio = mean(dspecc(nlow))-mean(dspecc(nhigh));
            para_EGF(nqsolve).specfit = dspecfit;
            para_EGF(nqsolve).ftest = ftest;
            para_EGF(nqsolve).etest = etest;
            para_EGF(nqsolve).exist = 1;
            para_EGF(nqsolve).fc = fc;
            para_EGF(nqsolve).fc2 = fc2;
            para_EGF(nqsolve).err = err;
            para_EGF(nqsolve).Es = Es;
            para_EGF(nqsolve).delsig = delsig;
            para_EGF(nqsolve).asig = asig;
            para_EGF(nqsolve).fcb1 = fcb1;
            para_EGF(nqsolve).fcb2 = fcb2;
            para_EGF(nqsolve).qid = evspec(iq).qid;
            para_EGF(nqsolve).qlat = evspec(iq).qlat;
            para_EGF(nqsolve).qlon = evspec(iq).qlon;
            para_EGF(nqsolve).qdep = evspec(iq).qdep;
            para_EGF(nqsolve).qmag = evspec(iq).qmagest;
            para_EGF(nqsolve).qmom = evspec(iq).qmomest;
            para_EGF(nqsolve).qmagcat = evspec(iq).qmag;
            para_EGF(nqsolve).qtime = evspec(iq).qtime;
            para_EGF(nqsolve).lhratio = lhratio;
            para_EGF(nqsolve).G = G;
            para_EGF(nqsolve).r = r;
            para_EGF(nqsolve).nspec = evspec(iq).nspec;

  	      if (mod(nqsolve,100) == 0)
  	      	fprintf(1,'event: %d, id: %d \n',nqsolve,evspec(iq).qid);
          end
          if (min(Es)>0 && iflag == 1 && fc<100)
              fprintf(fid4,format4,iq,evspec(iq).qmagest, evspec(iq).qmag,evspec(iq).qid,...
                  o0,fc,r,delsig,asig,...
                  G,Es(1:5),err, fcb1,fcb2);
              if (iexist == 1 & isavefig == 1 ) % just plot egf-stack fit
                  h1=figure(1);
                  set(h1,'visible','off');
                  subplot(2,2,1)
                  semilogx(freq,dspecc);
                  hold on
                  semilogx(freq,dspecfit,'-.');
                  hold off
                  xlabel('frequency');
                  xlim(frqlim);
                  ylabel('amplitude');

                  subplot(2,2,2);
                  plot(ftest,etest/min(etest));
                  ytest = min(etest)*1.05*[1 1]/min(etest);
                  hold on
                  plot([min(ftest) max(ftest)],ytest,'-.');
                  plot(fc,err/min(etest),'*');
                  hold off
                  xlabel('frequency');
                  ylabel('variance');
                  xlim([min(ftest) max(ftest)]);
                  title(strcat('shearer-stacking-',num2str(evspec(iq).qid)));
                  fileout = strcat(rdir,'fig-shearer-',num2str(evspec(iq).qid));
      			try
                    export_fig(fileout,'-pdf','-painters','-transparent');
    			catch
    				continue;
    			end
                close(1);
              end
          end
        end
    end
end

File: DESC_Python/src/specprocess_sourcepara.py
import os
import numpy as np
import matplotlib.pyplot as plt
from sourcepara import sourcepara


def specprocess_sourcepara(rdir, freqfit, evspec, imethod, iwave, EGF, min_nspec, isavefig):
    """
    Estimate source parameters using a constant EGF across events.

    Parameters
    ----------
    rdir : str
        Output directory.
    freqfit : tuple/list
        Frequency range for fitting (low, high).
    evspec : list of dict
        Event spectra with keys: 'freq', 'spec', 'qmagest', 'qmag', 'qid', 'qmomest',
        'qlat', 'qlon', 'qdep', 'qtime', 'nspec'.
    imethod : int
        Source fitting method (passed to sourcepara).
    iwave : int
        Wave type flag for sourcepara.
    EGF : ndarray
        Empirical Green's Function spectrum to subtract.
    min_nspec : int
        Minimum spectra count to include an event.
    isavefig : int/bool
        Whether to save diagnostic plots.

    Returns
    -------
    para_EGF : list of dict
        Fitted parameters per qualifying event.
    """

    if not os.path.isdir(rdir):
        os.makedirs(rdir, exist_ok=True)
    if not rdir.endswith('/'):
        rdir = rdir + '/'

    freq = np.array(evspec[0]['freq'])
    frqlim = [0.5, 50]
    flow = [0.5, 5]
    fhigh = [20, 30]
    nlow = np.where((freq >= flow[0]) & (freq <= flow[1]))[0]
    nhigh = np.where((freq >= fhigh[0]) & (freq <= fhigh[1]))[0]
    nfit = np.where((freq >= freqfit[0]) & (freq <= freqfit[1]))[0]

    out_path = os.path.join(rdir, 'result-constant-egf.txt')
    fmt_header = (
        "    i1   magest magcat              id1       o0             fc1"
        "          r    delsig    asig        G          Es-all   Es-model"
        "   Es-obs   Es-1       Es-2   error fcbound1 fcbound2\n"
    )
    fmt_line = (
        " %5d"  # i1
        "%8.2f"  # magest
        "%8.2f"  # magcat
        "%15d"   # id1
        "%15.2f" # o0
        "%15.2f" # fc1
        "%15.2e" # r
        "%10.3e" # delsig
        "%10.3e" # asig
        "%11.3e" # G
        "%10.3e" # Es-all
        "%10.3e" # Es-model
        "%10.3e" # Es-obs
        "%10.3e" # Es-1
        "%10.3e" # Es-2
        " %10.3e" # error
        " %10.3e" # fcb1
        " %10.3e" # fcb2
        "\n"
    )

    para_EGF = []
    nqsolve = 0

    with open(out_path, 'w') as fid4:
        fid4.write(fmt_header)

        for iq, ev in enumerate(evspec, start=1):
            if ev['nspec'] < min_nspec:
                continue

            dspecc = np.array(ev['spec']) - EGF
            if (dspecc.max() - dspecc.min()) <= 1e-2:
                continue

            nqsolve += 1
            fc, fc2, o0, Es, delsig, asig, r, G, err, dspecfit, iflag, ftest, etest, fcb1, fcb2, _ = \
                sourcepara(dspecc, freq, nfit, nlow, imethod, ev['qmomest'], iwave)

            lhratio = dspecc[nlow].mean() - dspecc[nhigh].mean()
            entry = {
                'specfit': dspecfit,
                'ftest': ftest,
                'etest': etest,
                'exist': 1,
                'fc': fc,
                'fc2': fc2,
                'err': err,
                'Es': Es,
                'delsig': delsig,
                'asig': asig,
                'fcb1': fcb1,
                'fcb2': fcb2,
                'qid': ev['qid'],
                'qlat': ev['qlat'],
                'qlon': ev['qlon'],
                'qdep': ev['qdep'],
                'qmag': ev['qmagest'],
                'qmom': ev['qmomest'],
                'qmagcat': ev['qmag'],
                'qtime': ev['qtime'],
                'lhratio': lhratio,
                'G': G,
                'r': r,
                'nspec': ev['nspec'],
            }
            para_EGF.append(entry)

            if nqsolve % 100 == 0:
                print(f"event: {nqsolve}, id: {ev['qid']}")

            if np.min(Es) > 0 and iflag == 1 and fc < 100:
                fid4.write(fmt_line % (
                    iq, ev['qmagest'], ev['qmag'], ev['qid'],
                    o0, fc, r, delsig, asig,
                    G, Es[0], Es[1], Es[2], Es[3], Es[4],
                    err, fcb1, fcb2
                ))

                if isavefig:
                    try:
                        plt.ioff()
                        fig = plt.figure(1, figsize=(8, 6))
                        plt.subplot(2, 2, 1)
                        plt.semilogx(freq, dspecc)
                        plt.semilogx(freq, dspecfit, '-.')
                        plt.xlabel('frequency')
                        plt.xlim(frqlim)
                        plt.ylabel('amplitude')

                        plt.subplot(2, 2, 2)
                        plt.plot(ftest, etest / np.min(etest))
                        ytest = np.min(etest) * 1.05 / np.min(etest)
                        plt.plot([ftest.min(), ftest.max()], [ytest, ytest], '-.')
                        plt.plot(fc, err / np.min(etest), '*')
                        plt.xlabel('frequency')
                        plt.ylabel('variance')
                        plt.xlim([ftest.min(), ftest.max()])
                        plt.title(f"shearer-stacking-{ev['qid']}")

                        fileout = os.path.join(rdir, f"fig-shearer-{ev['qid']}")
                        plt.savefig(fileout, format='pdf')
                        plt.close(fig)
                    except Exception:
                        # Skip plotting errors
                        pass

    return para_EGF
